catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/suggestion/MgrsCoordinateProcessor.java

```diff
@@ -0,0 +1,355 @@
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.ui.query.suggestion;
+
+import static java.lang.String.format;
+import static org.codice.ddf.catalog.ui.query.suggestion.LatLon.from;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.codice.ddf.spatial.geocoding.Suggestion;
+import org.codice.usng4j.BoundingBox;
+import org.codice.usng4j.CoordinateSystemTranslator;
+import org.codice.usng4j.UsngCoordinate;
+import org.codice.usng4j.impl.CoordinateSystemTranslatorImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A regex-based processor for identifying potential MGRS coordinate literals in a given input
+ * string despite deviations from the official spec. The general policy is to be as forgiving as
+ * possible without sacrificing determinism.
+ *
+ * <p>Due to the strict expectations of usng4j regarding input strings, some additional processing
+ * is done to "backfill" necessary zeros for occurrences of a {@link MgrsPartType#NUMERIC} string.
+ *
+ * @see #processNumericType(String) for more information on this "backfilling" behavior.
+ */
+@SuppressWarnings("squid:S1135" /* Ticket number provided on comment */)
+public class MgrsCoordinateProcessor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MgrsCoordinateProcessor.class);
+
+  private static final Pattern PATTERN_MGRS_ZONE = Pattern.compile("\\d\\d?[CDEFGHJKLMNPQRSTUVWX]");
+
+  private static final Pattern PATTERN_MGRS_100KM =
+      Pattern.compile("[ABCDEFGHJKLMNPQRSTUVWXYZ][ABCDEFGHJKLMNPQRSTUV]");
+
+  private static final Pattern PATTERN_MGRS_NUMERIC =
+      Pattern.compile("((\\d){1,5}(\\h)+(\\d){1,5}(\\h)+)|((\\d){2,10})");
+
+  private static final Pattern PATTERN_MGRS_ANYPART =
+      Pattern.compile(
+          format(
+              "(%s)|(%s)|(%s)",
+              PATTERN_MGRS_ZONE.pattern(),
+              PATTERN_MGRS_100KM.pattern(),
+              PATTERN_MGRS_NUMERIC.pattern()));
+
+  // usng4j helper constant for backfilling zeros
+  private static final Integer MAX_EASTING_LENGTH = 5;
+
+  // usng4j helper constant for backfilling zeros
+  private static final Integer MAX_NORTHING_LENGTH = 5;
+
+  // This key tells the UI that the geo is on the suggestion itself
```

Comment:
❓ Why do we need to be able to signify this? Shouldn’t there just be a method `getGeo` for a suggestion that handles this logic

Reply:
When would the geo not be a property of the suggestion?

Reply:
@bellcc _Why would we need to signify this?_

Normally the list of suggestions only contains objects with `id` and `name`. When a suggestion is picked, the `/geofeature?id=...` service is called to get the actual geometry. This saves the suggester endpoint valuable time NOT waiting for Solr. 

But with coordinate literals the geometry is already there (it can be inferred or computed from the input query) so the round trip to `/geofeature` is pointless. 

See https://github.com/codice/ddf/pull/3901#discussion_r227495806

But how should the UI know when to look for the geometry on the suggestion itself? There's probably a better way to do this, but for now, it can infer from the `id` field. Normally it would be a metacard ID ... a UUID. But if the `id` starts with "LITERAL" then we know that is not the case and that the `geo` attribute **should** be defined on the suggestion itself. 

Finally, since the `id` field needs to be unique (React keys on it within the child collection), we append an index. Currently, there will be at most 2 literal suggestions in the list, so for now this index can be hard-coded into the ID string. 

The `keyword.js` and `gazetteer.js` React components are worth a look to see how the UI handles things. 

@hayleynorton _When would the geo not be a property of the suggestion?_

We've defined our own impl of `Suggestion` to suit our needs but there's no guarantee that field lives on the impls being returned by the `FeatureService` call in `QueryApplication`.